source/common/upstream/upstream_impl.cc

```diff
@@ -1459,22 +1459,22 @@ bool BaseDynamicClusterImpl::updateDynamicHostList(const HostVector& new_hosts,
   const bool dont_remove_healthy_hosts =
       health_checker_ != nullptr && !info()->drainConnectionsOnHostRemoval();
   if (!current_priority_hosts.empty() && dont_remove_healthy_hosts) {
-    for (auto i = current_priority_hosts.begin(); i != current_priority_hosts.end();) {
-      if (!((*i)->healthFlagGet(Host::HealthFlag::FAILED_ACTIVE_HC) ||
-            (*i)->healthFlagGet(Host::HealthFlag::FAILED_EDS_HEALTH))) {
-        if ((*i)->weight() > max_host_weight) {
-          max_host_weight = (*i)->weight();
+    predicate = [&updated_hosts, &final_hosts, &max_host_weight](const HostSharedPtr& p) mutable {
+      if (!(p->healthFlagGet(Host::HealthFlag::FAILED_ACTIVE_HC) ||
+            p->healthFlagGet(Host::HealthFlag::FAILED_EDS_HEALTH))) {
+        if (p->weight() > max_host_weight) {
+          max_host_weight = p->weight();
         }
 
-        final_hosts.push_back(*i);
-        updated_hosts[(*i)->address()->asString()] = *i;
-        (*i)->healthFlagSet(Host::HealthFlag::PENDING_DYNAMIC_REMOVAL);
-        i = current_priority_hosts.erase(i);
-      } else {
-        i++;
+        final_hosts.push_back(p);
+        updated_hosts[p->address()->asString()] = p;
+        p->healthFlagSet(Host::HealthFlag::PENDING_DYNAMIC_REMOVAL);
+        return true;
       }
-    }
+      return false;
+    };
   }
+  Containers::removeMatchingElements(current_priority_hosts, predicate);
```

Comment:
should this be up one level, so it's inside the if-statement?

Reply:
yes!